APE10.rst

```diff
@@ -0,0 +1,144 @@
+Roadmap
+-------
+
+The proposed roadmap for releases, including past releases for reference, is the
+following::
+
+    v1.0 - LTS - February 19, 2015 (supported with bug fixes until June 2017)
+    v1.1 - December 11, 2015
+    v1.2 - June 23, 2016
+    v1.3 - December 2016
+    v2.0 - LTS - June 2017 (supported with bug fixes until end of 2019)
+    v3.0 - December 2017 - first release to support only Python 3+
+    v3.1 - June 2018
+    v3.2 - December 2018
+    v3.3 - June 2019
+    v4.0 - LTS - December 2019 (supported with bug fixes for two years)
+
+Note that v1.0, v2.0, and v4.0 are marked as long-term support releases (LTS),
+which means that they are typically supported with bug fixes for at least two
+years. The above roadmap proposes that the last Python 2-compatible release will
+be v2.0, which we will support with bug fixes until the end of 2019. In the mean
+time, v3.0, to be released in December 2017, will be the first release that does
+not support Python 2.
+
+Note that v3.0 would not be an LTS release, since we would
+otherwise need to maintain two LTS releases in parallel. This is a one-off
+deviation from the version numbering policy outlined in `APE 2
+<https://github.com/astropy/astropy-APEs/blob/master/APE2.rst>`_
+
+Implementation
+--------------
+
+In practice, dropping Python 2.7 support will involve, amongst other things:
+
+* Removing the bundled `six <https://pythonhosted.org/six/>`_ package, and
+  update all the code that used ``astropy.extern.six`` to use the Python 3
+  syntax directly.
+* Removing Python 2.7 from the configuration for the continuous integration
+  services (such as Travis and AppVeyor at the time of writing)
+* Making sure that the documentation on ReadTheDocs is built using Python 3.x
+* Removing any backports that may exist for Python 2.7 compatibility
+* Updating the documentation to remove any mention of Python 2.7, as well as
+  making sure that Python 3.x is not presented as a special case.
+* Remove Python 2 and 2.7 from the PyPI classifiers
+
+Challenges
+----------
+
+There are two main challenges with the plan outlined above:
+
+* Once we make the switch to a Python 3-only code base, between v2.0 and v3.0,
+  people will no longer be able to contribute code to Astropy without using
+  Python 3. This means that even though users can in theory wait until the end
+  of 2019 to switch to Python 3, developers and contributors will need to make
+  the switch (at least for their Astropy development environment) as soon as
+  2017.
+
+* If we upload Astropy v3.0 to PyPI, and it does not support Python 2, then if
+  Python 2.7 users try and pip install Astropy, the install or import will fail. This
+  is because PyPI does not compare the current Python version to the PyPI
+  meta-data for the package (which may indicate for example whether the package
+  is Python 2-compatible), and will download Astropy v3.0 regardless of the
+  active Python version. However, there is a solution, which is …
```

Comment:
https://github.com/jupyter/roadmap/blob/master/accepted/migration-to-python-3-only.md#multiple-source-distributions

```
Multiple source distributions

Alternatively, we could use a hidden feature of pip. Source distributions whose base name ends in -pyX.Y are considered for installation only on a matching Python version. Thus instead of publishing a single .tar.gz, we could publish multiple identical source distribution with various names.

For supported python versions:

ipython-6.x-py3.4.tar.gz
ipython-6.x-py3.5.tar.gz
For future python versions:

ipython-6.x-py3.6.tar.gz
ipython-6.x-py3.7.tar.gz
It may be possible to patch pip to handle a more general form like ipython-6.x-py3.tar.gz, so we only need to publish a single sdist.
```

Which is triggerd by https://github.com/pypa/pip/blob/a942fb9331762e88992c51fc3cf2ddffd5e4f3fb/pip/index.py#L635-L643

replacing `py_version != sys.version[:3]:` by `py_version != sys.version[:len(py_version)]:` should be enough. 

PyPI also support extra metadata [see https://pypi.python.org/pypi/flit] which says for example `Requires Python: 3` on PyPI UI, though it can't be set through setuptool, and is not supported by pip.